imgd.1
.TH "IMGD" "1" "Sep 2016" "Version 2.0" "User Commands"
.SH NAME
imgd \- Resize or rotate JPEG and PNG images.
.SH SYNOPSIS
.B imgd [OPTIONS] [PATH [PATH ...]]
.SH DESCRIPTION
.B imgd
(pronounced \fIimaged\fR) is a multiprocessing command line image resizer and rotator for JPEG and PNG images.
.PP
.B Features
.PP
  * resize by percentage or resolution
  * rotate clockwise by specified angle
  * adaptive resize considering orientation
  * brute force to a resolution
  * optimize images to save more space
  * convert PNG to JPEG
  * erase exif metadata
  * force smaller to larger resize
  * process directories recursively
  * overwrite source image option
.PP
Multiple files and directories can be specified as source. If \fBPATH\fR is omitted, the current directory is processed.
.PP
Output image names are appended with \fB_IMGD\fR if overwrite option is not used. By default \fB_IMGD\fR files are not processed. Doing so may lead to potential race conditions when overwrite option is used.
.PP
.B Adaptive mode
.PP
- If the specified and image orientations are same [(H >= V and h > v) or (H < V and h < v)], the image is resized with the longer specified side as reference.
.br
- In case of cross orientation [(H >= V and h <= v) or (H < V and h >= v)], the image is resized with the shorter specified side as reference. Same as non-adaptive.
.PP
For example, if an image has a resolution of 2048x1365 and is being resized to 1366x768:
.PP
- In regular mode (default), output image resolution will be 1152x768
.br
- In adaptive mode, output image resolution will be 1366x910
.SH OPTIONS
.TP
.BI "-h, --help"
Show help text and exit.
.TP
.BI "-s, --scale=" %
Scale image by percentage.
.TP
.BI "-x, --res=" HxV
Output resolution in HRESxVRES representation.
.TP
.BI "-o, --rotate=" deg
Rotate clockwise by a specified angle (in degrees).
.TP
.BI "-a, --adapt"
Adapt to specified resolution considering the orientation of the image. Disabled by default.
.TP
.BI "-c, --convert"
Convert PNG images to JPEG to save on space. The output image is saved with '.jpg' extension.
.TP
.BI "-d, --dot"
Include hidden files (Linux-specific). By default hidden files are skipped on Linux.
.TP
.BI "-e, --eraseexif"
Erase EXIF metadata of JPEG images. Preserved by default.
.TP
.B "-f, --force"
Force to the exact specified resolution. Disabled by default.
.TP
.B "-i, --includeimgd"
Process \fI_IMGD\fR files. Risky due to potential race conditions.
.TP
.B "-k, --keep"
Do not process if image hres or vres matches specified hres or vres, or scale is 100. However, PNG images are converted to JPEG if --convert option is specified.
.TP
.B "-n, --enlarge"
Enlarge smaller images. By default smaller images are not scaled if specified resolution is greater.
.TP
.BI "-p, --optimize"
Optimize output images using PIL library optimization algorithm. Disabled by default.
.TP
.BI "-q, --quiet"
Do not show any operational output.
.TP
.B "-r, --recursive"
Recursively process sub-directories. By default only the specified directory is processed. Symbolic links to directories are ignored to avoid recursive loops.
.TP
.BI "-w, --overwrite"
Overwrite the source images. By default an output image is saved with \fB_IMGD\fR appended to the source image name.
.br
.B NOTE:
If overwrite and convert options are used together, source PNG images are deleted.
.TP
.BI "-z, --debug"
Enable debugging.
.SH EXAMPLES
.PP
.IP 1. 4
Convert some images and directories:
.PP
.EX
.IP
.B $ imgd -x 1366x768 ~/ ~/Pictures/image3.png ~/Downloads/
/home/testuser/image1.png
3840x2160 -> 1365x768
11104999 bytes -> 1486426 bytes

/home/testuser/image2.jpg
2048x1365 -> 1152x768
224642 bytes -> 31421 bytes

/home/testuser/Pictures/image3.png
1920x1080 -> 1365x768
2811155 bytes -> 1657474 bytes

/home/testuser/Downloads/image4
2048x1365 -> 1152x768
224642 bytes -> 31421 bytes
.EE
.PP
.IP 2. 4
Scale an image by 75% and overwrite the source image:
.PP
.EX
.IP
.B $ imgd -s 75 -w ~/image.jpg
/home/testuser/image.jpg
1366x767 -> 1025x575
120968 bytes -> 45040 bytes
.EE
.PP
.IP 3. 4
Rotate an image clockwise by 90 degrees:
.PP
.EX
.IP
.B $ imgd -o 90 ~/image.jpg
120968 bytes -> 72038 bytes
.EE
.PP
.IP 4. 4
Adapt the images in the current directory to 1366x1000 resolution.
.br
Visit all directories recursively, overwrite source images, ignore images with matching hres or vres but convert PNG images to JPEG.
.PP
.EX
.IP
.B $ imgd -x 1366x1000 -acrw .
.EE
.PP
.EX
.IP 5. 4
Set hres=800 and adapt vres maintaining the ratio.
.PP
.EX
.IP
.B $ imgd -x 800x0
Source omitted. Processing current directory...

./image1.jpg
1366x911 -> 800x534
69022 bytes -> 35123 bytes

./image2.jpg
1050x1400 -> 800x1067
458092 bytes -> 78089 bytes
.SH AUTHORS
Arun Prakash Jana <engineerarun@gmail.com>
.SH HOME
.I https://github.com/jarun/imgd
.SH REPORTING BUGS
.I https://github.com/jarun/imgd/issues
.SH LICENSE
Copyright \(co 2016 Arun Prakash Jana <engineerarun@gmail.com>
.PP
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it. There is NO WARRANTY, to the extent permitted by law.


CHANGELOG
imgp v2.1
2016-10-05

Modifications
- Rename utility to `imgp`
- Support adapting to either hres or vres

-------------------------------------------------------------------------------

imgd v2.0
2016-09-30

News
- imgd is available on AUR

Modifications
- process current directory, if source unspecified
- completion scripts for bash, fish and zsh
- option `--processimgd` is now `--includeimgd`
- option `--keep` to skip images with matching hres/vres
- skip symbolic links to avoid recursive loops
- PEP8 compliance

-------------------------------------------------------------------------------

imgd v1.0
2016-08-17

Initial features
- resize by percentage or resolution
- rotate clockwise by specified angle
- adaptive resize considering orientation
- brute force to a resolution
- optimize images to save more space
- convert PNG to JPEG
- erase exif metadata
- force only larger to smaller resize
- process directories recursively
- overwrite source image option
- minimal dependencies

-------------------------------------------------------------------------------


imgd.py
#!/usr/bin/env python3
#
# Resize or rotate JPEG and PNG images.
#
# Copyright (C) 2016 Arun Prakash Jana <engineerarun@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import argparse
import sys
import os
import PIL
from PIL import Image, ImageFile
from multiprocessing import Pool, Lock


__version__ = '1.0'

HRES = 0
VRES = 0
scale = None
rotate = None
adapt = False
convert = False
dot = False
eraseexif = False
force = False
processimgd = False
keep = False
enlarge = False
optimize = False
quiet = False
recurse = False
overwrite = False
debug = False
pad = '_IMGD'
no_res_opt = False

pool = None
lock = Lock()

def getres(res):
    hxv = res.split('x')
    if len(hxv) != 2:
        return (0, 0)

    try:
        return(int(hxv[0]), int(hxv[1]))
    except:
        return (0, 0)

def lock_print(msg):
    global lock

    lock.acquire()
    print(msg)
    lock.release()

def traverse_dir(path):
    global lock, pool

    try:
        for entry in os.scandir(path):
            # Add check for hidden here
            if not dot and entry.name.startswith('.'):
                continue

            if entry.is_dir():
                if recurse:
                    if not os.access(entry.path, os.W_OK | os.X_OK):
                        lock_print(entry.path + ': no permission')
                        continue
                    traverse_dir(entry.path)
            elif rotate is not None:
                result = pool.apply_async(rotate_image, args=(entry.path,))
            else:
                result = pool.apply_async(resize_image, args=(entry.path,))
    except OSError as e:
        lock.acquire()
        print(e)
        lock.release()

def rotate_image(src):
    global lock

    converted = False
    name, ext = os.path.splitext(src)
    if name.endswith(pad) and not processimgd:
        return

    try:
        img = Image.open(src)
        _format = img.format
        if not (_format == 'JPEG' or _format == 'PNG'):
            if debug:
                lock_print(src + ': not JPEG or PNG image')
            return
        try:
            if eraseexif:
                exifdata = b''
            else:
                exifdata = img.info['exif']
        except:
            exifdata = b''
    except OSError as e:
        if not str(e).startswith('cannot identify image file'):
            lock.acquire()
            print(e)
            lock.release()
        return

    # Image.rotate(angle, resample=0, expand=0) expand = True means
    # the image will be enlarged to hold th entire rotated image
    img = img.rotate(-rotate, 0, True)

    if not quiet:
        stats = src + '\n%ld bytes ->' % os.path.getsize(src)

    if overwrite:
        suffix = ''
    else:
        suffix = pad

    if convert and _format == 'PNG':
        dest = name + suffix + '.jpg'
        _format = 'JPEG'
        converted = True
    else:
        dest = name + suffix + ext

    try:
        if not optimize:
            img.save(dest, _format, exif=exifdata)
        else:
            img.save(dest, _format, optimize=True, exif=exifdata)

        if not quiet:
            lock_print('%s %ld bytes\n' % (stats, os.path.getsize(dest)))

        if overwrite and converted:
            os.remove(src)
    except OSError as e:
        lock.acquire()
        print(e)
        lock.release()

def resize_image(src):
    global HRES, VRES, lock

    converted = False
    name, ext = os.path.splitext(src)
    if name.endswith(pad) and not processimgd:
        return

    try:
        img = Image.open(src)
        _format = img.format

        if not (_format == 'JPEG' or _format == 'PNG'):
            if debug:
                lock_print(src + ': not JPEG or PNG image')
            return

        # if 'only' convert option is specified,
        # nothing to be done for non-PNG images
        if no_res_opt and not optimize and _format != 'PNG':
            if debug:
                lock_print('resize_image: cannot convert non-PNG')
            return

        try:
            if eraseexif:
                exifdata = b''
            else:
                exifdata = img.info['exif']
        except:
            exifdata = b''
    except OSError as e:
        if not str(e).startswith('cannot identify image file'):
            lock.acquire()
            print(e)
            lock.release()
        return

    hres = img.size[0]
    vres = img.size[1]
    ImageFile.MAXBLOCK = hres * vres

    if keep and (HRES == hres or VRES == vres or scale == 100):
        if not (_format == 'PNG' and convert):
            return
    elif scale:
        if scale == 100:
            _hres = hres
            _vres = vres
        else:
            _hres = hres * scale/float(100)
            # use custom round()
            if _hres >= int(_hres) + .5:
                _hres = _hres + 1
            _hres = int(_hres)

            _vres = vres * scale/float(100)
            if _vres >= int(_vres) + .5:
                _vres = _vres + 1
            _vres = int(_vres)

        img = img.resize((_hres, _vres), PIL.Image.ANTIALIAS)
        if debug:
            lock_print('resize_image: scaled to [%dx%d]' % (HRES, VRES))
    elif HRES == hres and VRES == vres:
        if debug:
            lock_print('resize_image: same res [%dx%d]' % (HRES, VRES))
    elif force:
        img = img.resize((HRES, VRES), PIL.Image.ANTIALIAS)
        if debug:
            lock_print('resize_image: brute force res to [%dx%d]' % (HRES, VRES))
    elif adapt:
        ratio_img = float(hres/vres)
        ratio_target = float(HRES/VRES)

        if (ratio_target >= 1 and ratio_img > 1) or (ratio_target < 1 and ratio_img >= 1):
            # same orientation (H >= V and h > v) or
            # cross orientation (H < V and h >= v)
            if not enlarge and HRES > hres:
                return

            if HRES != hres:
                hratio = HRES/float(hres)
                target_vres = vres * hratio
                if target_vres >= int(target_vres) + .5:
                    target_vres = target_vres + 1
                target_vres = int(target_vres)

                img = img.resize((HRES, target_vres), PIL.Image.ANTIALIAS)
                if debug:
                    lock_print('resize_image: vres adapted %dx%d' % (HRES, target_vres))
        elif (ratio_target < 1 and ratio_img < 1) or (ratio_target >= 1 and ratio_img <= 1):
            # same orientation (H < V and h < v) or
            # cross orientation (H >= V and h <= v)
            if not enlarge and VRES > vres:
                return

            if VRES != vres:
                vratio = VRES/float(vres)
                target_hres = hres * vratio
                if target_hres >= int(target_hres) + .5:
                    target_hres = target_hres + 1
                target_hres = int(target_hres)

                img = img.resize((target_hres, VRES), PIL.Image.ANTIALIAS)
                if debug:
                    lock_print('resize_image: hres adapted %dx%d' % (target_hres, VRES))
    else:
        ratio_img = float(hres/vres)
        ratio_target = float(HRES/VRES)

        if ratio_img >= ratio_target:
            if not enlarge and HRES > hres:
                return

            # re-sample as per target horizontal resolution
            hratio = HRES/float(hres)
            target_vres = vres * hratio
            if target_vres >= int(target_vres) + .5:
                target_vres = target_vres + 1
            target_vres = int(target_vres)

            img = img.resize((HRES, target_vres), PIL.Image.ANTIALIAS)
            if debug:
                lock_print('resize_image: vres calculated %dx%d' % (HRES, target_vres))
        else:
            if not enlarge and VRES > vres:
                return

            # re-sample as per target vertical resolution
            vratio = VRES/float(vres)
            target_hres = hres * vratio
            if target_hres >= int(target_hres) + .5:
                target_hres = target_hres + 1
            target_hres = int(target_hres)

            img = img.resize((target_hres, VRES), PIL.Image.ANTIALIAS)
            if debug:
                lock_print('resize_image: hres calculated %dx%d' % (target_hres, VRES))

    if not quiet:
        stats = src + '\n%dx%d -> %dx%d\n%ld bytes ->' \
            % (hres, vres, img.size[0], img.size[1], os.path.getsize(src))

    if overwrite:
        suffix = ''
    else:
        suffix = pad

    if convert and _format == 'PNG':
        dest = name + suffix + '.jpg'
        _format = 'JPEG'
        converted = True
    else:
        dest = name + suffix + ext

    try:
        if not optimize:
            img.save(dest, _format, exif=exifdata)
        else:
            img.save(dest, _format, optimize=True, exif=exifdata)

        if not quiet:
            lock_print('%s %ld bytes\n' % (stats, os.path.getsize(dest)))

        if overwrite and converted:
            os.remove(src)
    except OSError as e:
        lock.acquire()
        print(e)
        lock.release()

class ExtendedArgumentParser(argparse.ArgumentParser):
    """Extend classic argument parser"""

    # Print additional help and info
    @staticmethod
    def print_extended_help(file=None):
        if not file:
            file = sys.stderr

        file.write('''
Version %s
Copyright (C) 2016 Arun Prakash Jana <engineerarun@gmail.com>
License: GPLv3
Webpage: https://github.com/jarun/imgd
''' % __version__)

    # Help
    def print_help(self, file=None):
        super(ExtendedArgumentParser, self).print_help(file)
        self.print_extended_help(file)

def parse_args(args=None, namespace=None):
    """Parse imgd arguments/options.
    Parameters
    ----------
    args : list, optional
        Arguments to parse. Default is ``sys.argv``.
    namespace : argparse.Namespace
        Namespace to write to. Default is a new namespace.
    Returns
    -------
    argparse.Namespace
        Namespace with parsed arguments / options.
    """

    argparser = ExtendedArgumentParser(description='Resize or rotate JPEG and PNG images.')
    addarg = argparser.add_argument
    addarg('-s', '--scale', dest='scale', type=int, metavar='%',
           help='scale image by percentage')
    addarg('-x', '--res', dest='res', nargs=1, metavar='HxV',
           help='output resolution in HxV representation')
    addarg('-o', '--rotate', dest='rotate', type=int, metavar='deg',
           help='rotate clockwise by specified angle')
    addarg('-a', '--adapt', dest='adapt', action='store_true',
           help='adapt to resolution by orientation [default: off]')
    addarg('-c', '--convert', dest='convert', action='store_true',
           help='convert PNG to JPG format [default: off]')
    addarg('-d', '--dot', dest='dot', action='store_true',
           help='include hidden files (on Linux) [default: off]')
    addarg('-e', '--eraseexif', action='store_true',
           help='erase exif metadata [default: off]')
    addarg('-f', '--force', dest='force', action='store_true',
           help='force to exact specified resolution [default: off]')
    addarg('-i', '--processimgd', dest='processimgd', action='store_true',
           help='re-process generated files. * RISKY: refer to docs')
    addarg('-k', '--keep', dest='keep', action='store_true',
           help='skip (but convert, if opted) images with matching specified \
                 hres or vres or scale=100 [default: off]')
    addarg('-n', '--enlarge', dest='enlarge', action='store_true',
           help='enlarge smaller images [default: off]')
    addarg('-p', '--optimize', dest='optimize', action='store_true',
           help='optimize the output images [default: off]')
    addarg('-q', '--quiet', dest='quiet', action='store_true',
           help='operate silently [default: informative]')
    addarg('-r', '--recurse', dest='recurse', action='store_true',
           help='process directories recursively [default: off]')
    addarg('-w', '--overwrite', dest='overwrite', action='store_true',
           help='overwrite source images [default: off]')
    addarg('-z', '--debug', dest='debug', action='store_true',
           help='enable debug logs [default: off]')
    addarg('keywords', nargs='+', metavar='PATH',
           help='source file or directory')

    # Show help and exit if no arguments
    if len(sys.argv) < 2:
        argparser.print_help(sys.stderr)
        sys.exit(1)

    return argparser.parse_args(args, namespace)

def main():
    global HRES, VRES, scale, rotate, adapt, convert, dot, eraseexif, \
           force, processimgd, keep, enlarge, optimize, \
           quiet, recurse, overwrite, debug, no_res_opt, pool

    args = parse_args()

    if (args.res is not None and args.scale is not None):
        print('specify either resolution or scale')
        return

    if (args.rotate is not None and (args.res is not None or args.scale is not None)):
        print('option --rotate cannot be used with --scale or --res')
        return

    if args.scale is not None:
        if (args.scale <= 0):
            print('scale should be > 0%')
            return
        if args.scale > 100 and not args.enlarge:
            print('use --enlarge is to scale > 100%')
            return
        scale = args.scale
    elif args.res is not None:
        HRES, VRES = getres(args.res[0])
        if HRES == 0 or VRES == 0:
            print('resolution must be numeric and > 0')
            return
    elif args.rotate is not None:
        if(args.rotate <= 0) or (args.rotate >= 360):
            print('the angle specified should be > 0 degree and < 360 degree')
            return
        rotate = args.rotate
    elif args.convert or args.optimize:
        no_res_opt = True
        scale = 100
    else:
        print('missing image transformation')
        return

    if len(args.keywords) == 0:
        print('no source specified')
        return

    adapt = args.adapt
    convert = args.convert
    dot = args.dot
    eraseexif = args.eraseexif
    force = args.force
    processimgd = args.processimgd
    keep = args.keep
    enlarge = args.enlarge
    optimize = args.optimize
    quiet = args.quiet
    recurse = args.recurse
    overwrite = args.overwrite
    debug = args.debug

    pool = Pool()

    for path in args.keywords:
        if not os.path.exists(path):
            print('%s does not exist' % path)
        elif os.path.isdir(path):
            traverse_dir(path)
        elif args.rotate is not None:
            rotate_image(path)
        else:
            resize_image(path)

    pool.close()
    pool.join()

if __name__ == '__main__':
    main()
